### PR TITLE
netcdf: multiple improvements

### PR DESCRIPTION
**Pull request by aprokop**
- Added 'dap' and 'cdmremote' variants
  This is based on work in #2324 with the following motivation:
      Turn off DAP support by deafult.  DAP requires curl, which has issues
      with circular dependencies.  For 95% of NetCDF users that do not need
      DAP, turning it off avoides this rats nest of problems.
- Added 'parallel-netcdf' variant
  To support work with parallel-netcdf
- Added 'shared' and 'static' build separation

### CODE REVIEW

File: var/spack/repos/builtin/packages/netcdf/package.py [hunk 1]
```diff
@@ -99,17 +104,36 @@ def install(self, spec, prefix):
...
+            config_args.append('--disable-dap')
+
+        if '+cdmremote' in spec:
+            config_args.append('--enable-cdmremote')
+        else:
+            config_args.append('--disable-cdmremote')
+
+        if '+dap' in spec or '+cdmremote' in spec:
+            # Make sure Netcdf links against Spack's curl, otherwise
+            # otherwise it may pick up system's curl, which can give link
```
Comment (davydden) [inline var/spack/repos/builtin/packages/netcdf/package.py]: two `otherwise`

File: var/spack/repos/builtin/packages/netcdf/package.py [hunk 2]
```diff
@@ -39,8 +39,13 @@ class Netcdf(Package):
     version('4.4.0', 'cffda0cbd97fdb3a06e9274f7aef438e')
     version('4.3.3', '5fbd0e108a54bd82cb5702a73f56d2ae')
 
-    variant('mpi',  default=True,  description='Enables MPI parallelism')
-    variant('hdf4', default=False, description='Enable HDF4 support')
+    variant('mpi',     default=True,  description='Enables MPI parallelism')
+    variant('hdf4',    default=False, description='Enable HDF4 support')
+    variant('shared',  default=True,  description='Enable shared library')
+    variant('static',  default=True,  description='Enable static library')
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/netcdf/package.py]: This is the only package that uses the `static` variant, and if I am not wrong we decided a while ago to use just `shared`. Anyhow: what would be the semantic of
```console
spack install netcdf~shared~static
```
that is now allowed?

Comment (aprokop) [inline var/spack/repos/builtin/packages/netcdf/package.py]: @davydden @alalazo Is an overall Spack guidance for packages for shared/static builds documented somewhere?

Comment (alalazo) [inline var/spack/repos/builtin/packages/netcdf/package.py]: @aprokop I thought it was, but sadly I cannot find it. What all the other packages do is to build both shared and static libraries when `+shared` and just static when `~shared`.

Comment (citibeth) [inline var/spack/repos/builtin/packages/netcdf/package.py]: See here:

https://groups.google.com/forum/#!topic/spack/gNAp1CmhxF0

**Comment by adamjstewart** [comment]
@citibeth If you download the latest version, 4.1.1.1, it does indeed have the `--enable-pnetcdf` flag available. I didn't know that it depended on `parallel-netcdf` though.

**Comment by citibeth** [comment]
> @citibeth If you download the latest version, 4.1.1.1, it does indeed have the --enable-pnetcdf flag available. I didn't know that it depended on parallel-netcdf though.

Thanks, good detective work.  Like you, without further evidence I'm still dubious that `netcdf` 4 would depend on `parallel-netcdf`.

**Comment by aprokop** [comment]
> I'm concerned about depending on parallel-netcdf. What version of NetCDF are you building here? The latest version of NetCDF4 does not seem to have a configure flag named --enable-pnetcdf.

Oh, I did not even realize that. I'm building netcdf 4.3.3.1, and was told by a person of the code that it needs both `--enable-pnetcdf` and `–enable-netcdf-4`.

I will try to dig deeper to figure out when pnetcdf option is available.

**Comment by aprokop** [comment]
> @adamjstewart  If you download the latest version, 4.1.1.1
Did you mean 4.4.1.1?

According to [this](http://trac.mcs.anl.gov/projects/parallel-netcdf/wiki/PnetcdfAndNetcdf4), the `--enable-pnetcdf` option appeared in 4.2.2.1.

Couple quick questions:
1. Is one supposed to hide a spec variant if it does not do anything for a particular spec version?
2. Is it possible to combine statement in `when` clause? Something like this:
```python
    depends_on("parallel-netcdf", when='+parallel-netcdf' && '@4.2.1.1:')
```
Or do I have to do the `if`?

**Comment by aprokop** [comment]
Updated PR:
- Got rid of `static` variant
  The current spec always enables static
- If `~shared`, add `-fPIC` flag
   This would have to be changed when #2375 is merged
- Update the requirement for `parallel-netcdf` variant 
- Fixed double `otherwise` misprint

I had to add `-fPIC` to CFLAGS instead of CPPFLAGS as the netcdf configure has this:
```
  CFLAGS      C compiler flags
  LDFLAGS     linker flags, e.g. -L<lib dir> if you have libraries in a
              nonstandard directory <lib dir>
  LIBS        libraries to pass to the linker, e.g. -l<library>
  CPPFLAGS    (Objective) C/C++ preprocessor flags, e.g. -I<include dir> if
              you have headers in a nonstandard directory <include dir>
```

**Comment by aprokop** [comment]
Updated PR with @adamjstewart  comments on `when`.

File: var/spack/repos/builtin/packages/netcdf/package.py [hunk 3]
```diff
@@ -131,10 +151,18 @@ def install(self, spec, prefix):
             LDFLAGS.append("-L%s/lib"     % spec['szip'].prefix)
             LIBS.append("-l%s"         % "sz")
 
+        # PnetCDF support
+        if '+parallel-netcdf' in spec:
+            config_args.append('--enable-pnetcdf')
+            config_args.append('CC=%s' % spec['mpi'].mpicc)
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/netcdf/package.py]: What happens when someone tries to build `netcdf+parallel-netcdf~mpi`?

Comment (adamjstewart) [inline var/spack/repos/builtin/packages/netcdf/package.py]: Also, out of curiosity, what exactly is the difference between `netcdf+mpi` and `netcdf+parallel-netcdf`? I've never used either, but my users regularly use serial `netcdf` and `parallel-netcdf`.

Comment (davydden) [inline var/spack/repos/builtin/packages/netcdf/package.py]: > The simplest way would be to prohibit parallel-netcdf without mpi

i would say thrown an error if someone tries to do it.

**Comment by citibeth** [comment]
Something in me is not comfortable with a variant having a dash in it.
Could we rename `+parallel-netcdf` --> `+pnetcdf`?  Would also rename the
package similarly.   This would not be inappropriate, since the library
calls itself both:

https://trac.mcs.anl.gov/projects/parallel-netcdf

 The difference is... `pnetcdf` is a reimplementation of the NetCDF-3 file
format for parallel.  It is from Argonne National Lab and completely
different and separate from the NetCDF effort at Unidata.

NetCDF4 is no longer a format itself; it is a front-end on top of an HDF5
backend --- and a NetCDF3 backend for compatibility.  The recent change
pointed to above allows one to use `pnetcdf` as a backend as well.

If building a new project, I would use NetCDF4+HDF5, since HDF5 also offers
parallel capabilities, and its file format has a lot of useful features not
found in the NetCDF3 format supported by `pnetcdf`.

**Comment by citibeth** [comment]
Does the `--enable-pnetcdf` option require MPI directly?  Meaning... does
NetCDF make MPI calls directly in that case?  Or does it just use MPI
through `pnetcdf`, in which case we don't have to do anything?

**Comment by adamjstewart** [comment]
I agree with renaming the variant to `pnetcdf`, since that's what the actual configure flag looks like. As for the name of the `parallel-netcdf` package, I'm indifferent.

**Comment by aprokop** [comment]
> Does the `--enable-pnetcdf` option require MPI directly?  Meaning... does
> NetCDF make MPI calls directly in that case?  Or does it just use MPI
> through `pnetcdf`, in which case we don't have to do anything?

I'm actually not sure any more. Looking at the old spec, `+mpi` adds `--enable-parallel4` config option, but for instance 4.3.3.1 does not have it (I think it only appeared in 4.4).

Few interesting bits of information from [here](http://www.unidata.ucar.edu/software/netcdf/docs/getting_and_building_netcdf.html):
- The dependencies of netcdf are only these:
   - HDF5 1.8.9 or later (for netCDF-4 support)
   - zlib 1.2.5 or later (for netCDF-4 compression)
   - curl 7.18.0 or later (for DAP remote access client support)
-  But in the [Building with parallel support](http://www.unidata.ucar.edu/software/netcdf/docs/getting_and_building_netcdf.html#build_parallel) section it says to use `mpicc` to compile netcdf when used with `pnetcdf` or with parallel HDF5

So, I think it does need MPI when compiling with `+pnetcdf` or with `^hdf5+mpi`.  I'm less sure now of whether I need the
```
config_args.append('CC=%s' % spec['mpi'].mpicc)
```
line as I only needed that for static builds, and I have not checked for shared.

In summary, a) I will rename the variant to `pnetcdf`; b) I'll make sure `+pnetcdf` requires `+mpi` and leave it at that; and c) I will also check whether it works without explicit `mpicc`. 

Regarding the `parallel-netcdf` package renaming, I have no opinion.

**Comment by citibeth** [comment]
Consider using the netcdf cmake build instead of autotools.  Cmake knows
how to set pic flags for different compilers.

**Comment by aprokop** [comment]
> Consider using the netcdf cmake build instead of autotools.  Cmake knows
> how to set pic flags for different compilers.

Would not that require to completely redo the whole spec?

**Comment by adamjstewart** [comment]
We've discussed CMake vs. Autotools for the NetCDF package before: https://github.com/LLNL/spack/pull/400#issuecomment-175353103

We decided to stick with Autotools because it can build older versions of Autotools. I also don't see any problems with the current Autotools build.

**Comment by davydden** [comment]
I would keep 'parallel-netcdf' as package name as it is self-explanatory, whereas pnetcdf to me sounds close to python namings (i.e. py-netcdf).

**Comment by tgamblin** [comment]
@aprokop @citibeth: how about `parallel` for the variant name?  We've used this on some other packages.  I'm sometimes not a fan of it because `parallel` can mean MPI, OpenMP, pthreads, or whatever, but there is really only one type of parallel NetCDF and I think the concept of parallel I/O, at least in HPC, is fairly well understood.  Decent compromise?

**Comment by citibeth** [comment]
I would rather just keep `parallel-netcdf`

**Comment by tgamblin** [comment]
Going with `parallel-netcdf` as I'm leaning more towards that now that I see it's the same name as the dependency (so I think it is clearer).  If people feel strongly please submit a new PR and we can debate the name.
